### PR TITLE
Add FixReturnType fixer

### PR DESCRIPTION
**Pull request by jcouv**
If you have a method that returns `void` (or an async method that returns `void` or `Task`), but you have a return statement with a value, the FixReturnType fixer offers to fix the return type.

image

Update: I'll extend this PR to also handle VB.

### CODE REVIEW

**Comment by dotMorten** [comment]
Yes so much this! Quite often as you write the method you realize the return type should be something different.

**Comment by ghuntley** [comment]
Amaze turn around.

File: src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+        [Fact]
+        public async Task ReturnNull()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class C
+{
+    void M()
+    {
+        [|return|] null;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs]: perhaps offer to just be 'object'?  we often infer that as the 'default' type to use if we can't figure out something better.

File: src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    async void M()
+    {
+        [|return|] """";
+    }
+}",
+@"class C
+{
+    async System.Threading.Tasks.Task<string> M()
+    {
+        return """";
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs]: is there a test that shows this is simplified if the right using is in place?

File: src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        async System.Threading.Tasks.Task<string> local()
+        {
+            return """";
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs]: can we have tests with trivia, validating that updating the return type doesn't mess up trivia?

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CS0127, CS1997);
+
+        public CSharpFixReturnTypeCodeFixProvider()
+            : base(supportsFixAll: false)
+        {
+        }
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            if (await TryGetOldAndNewReturnTypeAsync(context.Document, context.Diagnostics, context.CancellationToken).ConfigureAwait(false) == default)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: 1. so long.  and hard to see the `== default`.  can you break into separate statements?
2. consider locals for 'document/diagnostics'.

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+               new MyCodeAction(c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+               context.Diagnostics);
+        }
+
+        private async Task<(TypeSyntax declarationToFix, TypeSyntax fixedDeclaration)> TryGetOldAndNewReturnTypeAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            Debug.Assert(diagnostics.Length == 1);
+            var location = diagnostics[0].Location;
+            var returnStatement = (ReturnStatementSyntax)location.FindNode(getInnermostNodeForTie: true, cancellationToken: cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: second named param necessary?

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var returnedType = semanticModel.GetTypeInfo(returnedValue, cancellationToken).Type;
+            if (returnedType == null)
+            {
+                return default;
+            }
+
+            var fixedType = useTask
+                ? semanticModel.Compilation.TaskOfTType().Construct(returnedType)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: TaskOfTType may not be available.  this will throw.

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            return (declarationTypeToFix, fixedDeclaration);
+        }
+
+        protected override async Task FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var (declarationTypeToFix, fixedDeclaration) =
+                await TryGetOldAndNewReturnTypeAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
+
+            editor.ReplaceNode(declarationTypeToFix, fixedDeclaration);
+            return;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: unnecessary.

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            if (!node.IsKind(SyntaxKind.ReturnStatement))
+            {
+                return default;
+            }
+
+            var containingMember = node.GetAncestors().FirstOrDefault(a => a.IsKind(
+                SyntaxKind.MethodDeclaration, SyntaxKind.PropertyDeclaration, SyntaxKind.ParenthesizedLambdaExpression, SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.LocalFunctionStatement, SyntaxKind.AnonymousMethodExpression, SyntaxKind.ConstructorDeclaration, SyntaxKind.DestructorDeclaration,
+                SyntaxKind.OperatorDeclaration, SyntaxKind.IndexerDeclaration, SyntaxKind.EventDeclaration));
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: ick.  feels like there's something in the IDE somewhere for this :)

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            // Local functions
+            bool IsAsync(SyntaxTokenList modifiers)
+            {
+                return modifiers.Any(SyntaxKind.AsyncKeyword);
+            }
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument) :
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: `:` on next line.

**Comment by CyrusNajmabadi** [comment]
Two design points (can come later/never):
1. C# only?
2. Only for fixing up 'void' return types?

**Comment by jcouv** [comment]
> * C# only?

I'll extend this PR to also handle VB.

> * Only for fixing up 'void' return types?

Yes, that's an intentional limitation at the moment. I plan to leave as-is.

File: src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs (hunk 3 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs]: The trivia comes out with an extra newline. That comes from formatting (ie. I think it's an existing issue), so I'll leave it as-is for now.

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs (hunk 6 above)
Comment (jcouv) [reply src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: Indeed. Thanks

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs (hunk 8 above)
Comment (jcouv) [reply src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: Looked around, but didn't find anything good. Ended up removing it.

**Comment by alrz** [comment]
Tracking issues: https://github.com/dotnet/roslyn/issues/31969, https://github.com/dotnet/roslyn/issues/29677

**Comment by jcouv** [comment]
Thanks @alrz for the links. In particular, here are the [design notes](https://github.com/dotnet/roslyn/issues/29677#issuecomment-426251558).
Also, @MaStr11 documented a [list of scenarios](https://github.com/dotnet/roslyn/issues/29677#issuecomment-426630329) to consider. I'm still mostly focused on `void` scenario for now, but he provided some good ideas (for example, fix `void M() => 1;`).

File: src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+@"class C
+{
+    void M()
+    {
+        [|return|] () => {};
+    }
+}",
+ @"class C
+{
+    object M()
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs]: Is this really helpful? Shouldn't we deduce the actual type here such as `Action`, `Func`, etc. If you feel that is out of scope of the PR, probably worth filing a tracking work item.

File: src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+               new MyCodeAction(c => FixAsync(document, diagnostics.First(), c)),
+               diagnostics);
+        }
+
+        private async Task<(TypeSyntax declarationToFix, TypeSyntax fixedDeclaration)> TryGetOldAndNewReturnTypeAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            Debug.Assert(diagnostics.Length == 1);
+            var location = diagnostics[0].Location;
+            var returnStatement = (ReturnStatementSyntax)location.FindNode(getInnermostNodeForTie: true, cancellationToken);
```
Comment (mavasani) [inline src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: Just curious: what diagnostic fires for expression bodied members declared as returning void?

Comment (jcouv) [reply src/Features/CSharp/Portable/CodeFixes/FixReturnType/CSharpFixReturnTypeCodeFixProvider.cs]: Added a test and support for that scenario.

File: src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs (hunk 10 above)
Comment (jcouv) [reply src/EditorFeatures/CSharpTest/Diagnostics/FixReturnType/FixReturnTypeTests.cs]: I agree it's not great. I'll file a follow-up issue.

**Comment by jcouv** [comment]
@mavasani FYI I'll keep VB support for a follow-up PR. Let me know if that's not good.

**Comment by jcouv** [comment]
Thanks! Merged.
